lib/ddtrace/profiling.rb

```diff
@@ -4,24 +4,61 @@ module Profiling
     module_function
 
     GOOGLE_PROTOBUF_MINIMUM_VERSION = Gem::Version.new('3.0')
+    private_constant :GOOGLE_PROTOBUF_MINIMUM_VERSION
 
     def supported?
-      google_protobuf_supported?
+      unsupported_reason.nil?
     end
 
-    def native_cpu_time_supported?
-      require 'ddtrace/profiling/ext/cpu'
-      Ext::CPU.supported?
+    def unsupported_reason
+      # NOTE: Only the first matching reason is returned, so try to keep a nice order on reasons -- e.g. tell users
+      # first that they can't use this on JRuby before telling them that they are missing protobuf
+
+      if RUBY_ENGINE == 'jruby'
+        'JRuby is not supported'
+      elsif Gem::Version.new(RUBY_VERSION) < Gem::Version.new('2.1')
+        'Ruby >= 2.1 is required'
+      elsif Gem.loaded_specs['google-protobuf'].nil?
+        "Missing google-protobuf dependency; please add `gem 'google-protobuf', '~> 3.0'` to your Gemfile or gems.rb file"
+      elsif Gem.loaded_specs['google-protobuf'].version < GOOGLE_PROTOBUF_MINIMUM_VERSION
+        'Your google-protobuf is too old; ensure that you have google-protobuf >= 3.0 by ' \
+        "adding `gem 'google-protobuf', '~> 3.0'` to your Gemfile or gems.rb file"
+      elsif !protobuf_loaded_successfully?
+        'There was an error loading the google-protobuf library; see previous warning message for details'
+      end
     end
 
-    def google_protobuf_supported?
-      RUBY_PLATFORM != 'java' \
-        && !Gem.loaded_specs['google-protobuf'].nil? \
-        && Gem.loaded_specs['google-protobuf'].version >= GOOGLE_PROTOBUF_MINIMUM_VERSION \
-        && !defined?(@failed_to_load_protobuf)
+    # The `google-protobuf` gem depends on a native component, and its creators helpfully tried to provide precompiled
+    # versions of this extension on rubygems.org.
+    #
+    # Unfortunately, for a long time, the supported Ruby versions metadata on these precompiled versions of the extension
+    # was not correctly set. (This is fixed in newer versions -- but not all Ruby versions we want to support can use
+    # these.)
+    #
+    # Thus, the gem can still be installed, but can be in a broken state. To avoid breaking customer applications, we
+    # use this helper to load it and gracefully handle failures.
+    def protobuf_loaded_successfully?
```

Comment:
Not sure if this needs to be "public". I think it's an internal implementation detail, for which we are introducing clear error messaging when the user has to take action on it.
But I'll let this decision to you, not blocking.